llext-edk: capture full build properties for EDK export

pillo79 opened the pull request:
### The symptom

While developing the Arduino core, we encountered an issue compiling code that used the `fatfs` module via the EDK; it looked like some defaults were being used in the extension, even though specific compile options were applied during the main Zephyr build.

### The actual problem

Tracking down the source revealed that the EDK is not exporting all build information, but only those defined by the Zephyr core (the `zephyr_interface` target). Modules do not interfere with the Zephyr core but insert themselves as dependencies of the `app` target, so transitive evaluation applies there only. 

### Ruled-out solutions

* The current way of reading properties (e.g. `zephyr_get_compile_options_for_lang`) cannot be used for the `app` target, because at the time the main Zephyr `CMakeLists.txt` is evaluated, that target is not yet available (and definitely not fully configured).

* Trying to use generator expressions is also not possible, due to CMake not allowing expansion of `$<COMPILE_LANGUAGE:...>` outside of a proper compile context. These expressions were stripped by complex code in `get_*_for_lang` functions, but that's no longer possible if the actual flags are hidden in a genex. The process errors out at configuration time.

### The proposed solution

To work around this, this change introduces a new CMake subdirectory (`misc/llext_edk`) that creates a dummy target which imports the build configuration from `app`, but overrides the C compile rules to simply output the resulting properties (defines, include directories, flags) to text files. The EDK generation script then reads those text files to get the full set of build properties, and includes them in the EDK tarball.

> [!IMPORTANT]
>  This (ab-)uses the CMake build process in a way that is often cited as an error - setting toolchain variables from a `CMakeLists.txt` selectively overrides an existing configuration, and is typically discouraged. However, that's 1) documented, stable behavior and 2) exactly what we are after in this case: use every option from the existing toolchain, but "discard" the compiler executable.

A test has been added to verify the functionality by checking if a symbol defined in an interface library that is a dependency of `app` is properly exported to the extension.

This has been in use for the past 4 months in the Arduino core for Zephyr workflows and it has fixed numerous issues related to missing compile options.

pillo79 committed:
llext-edk: add dependency flag testing

Add an interface library to the test application that defines a macro
whose value is later used in the extension. This verifies the full
dependency chain for 'app' is evaluated during EDK generation.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext-edk: (refact) filter flags in post-processing step

This changes the LLEXT EDK flag filtering to be done after the main
Zephyr build is completed. Instead of filtering the flags via genexes,
the full set is passed to the script along with the set of filters to be
applied, and the filter is applied later.

No functional change is intended.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext-edk: capture full build properties for EDK export

The current implementation of the LLEXT EDK exports only a partial set
of compiler settings to the EDK build system, specifically only the ones
defined by the Zephyr core (the 'zephyr_interface' target). Extending
this to include application and module-specific settings (from the 'app'
target and its dependencies) is tricky, because:
- while evaluating the main CMakeLists.txt, the 'app' target is not
  fully configured yet, and
- using generator expressions is not possible due to CMake not allowing
  expansion of $<COMPILE_LANGUAGE:...> outside of a proper compile
  context.

To work around this, this change introduces a new CMake subdirectory
(misc/llext_edk) that creates a dummy target which imports the build
configuration from 'app', but overrides the C compile rules to simply
output the resulting properties (defines, include directories, flags)
to text files. The EDK generation script then reads those text files
to get the full set of build properties, and includes them in the EDK
tarball.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

edersondisouza reviewed (commented):
Interesting hack! Two questions:

 1- I noticed that there seems to be less duplication of flags - like, less times the drivers include dir appears on the flags after your changes - is this expected? Why did we have some many duplicates anyway?
 2 - Where is this documented as a stable behaviour? Would it make sense to have a link to that for future reference?

tejlmand reviewed (dismissed):
Looks good, but a minor improvement proposal on the comment text.

pillo79 committed:
llext-edk: remove 'llext_filter_zephyr_flags' helper function

This function was introduced when the EDK was first implemented to reuse
the flag filtering code in 'add_llext_target'.

After moving the EDK filtering code to a later stage where simpler regex
eplacements can be used, the helper function is used only in its former
location and is no longer needed. Remove the otherwise-unused public
function to simplify the codebase.

This is a revert of commit dc8d7ada199aac3d204d35a3805c17beb721a5ac
("cmake/modules: Extract flags filtering code to a function") with
minimal code style cleanups.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

tmleman commented:
This PR breaks SOF compilation in Windows environment on platforms with LLEXT enabled. The problem is caused by commit a97b5f221bf178c3427d3031e8f2a2d5ca922644.

Link to SOF workflow: https://github.com/thesofproject/sof/actions/runs/25675119170/job/75374858061?pr=10765
